src/tlog/tlc2.ml

```diff
@@ -725,164 +472,159 @@ class tlc2
+    method which_tlog_file (n:int) =
+      TlogMap.which_tlog_file tlog_map n
+
 
 
-    method dump_tlog_file start_i oc =
-      self # which_tlog_file start_i >>= fun co ->
+    method dump_tlog_file n oc =
+      self # which_tlog_file n >>= fun co ->
       let canonical = match co with
         | Some fn -> fn
-        | None -> let n = Sn.to_int (get_file_number start_i) in
+        | None -> 
                   let fn = file_name n in
-                  get_full_path tlog_dir tlf_dir fn
+                  TlogMap.get_full_path tlog_map fn
       in
-      Logger.debug_f_ "start_i = %Li => canonical=%s" start_i canonical >>= fun () ->
+      Logger.debug_f_ "n = %03i => canonical=%s" n canonical >>= fun () ->
       let name = Filename.basename canonical in
-      Llio.output_string oc name >>= fun () ->
-      File_system.stat canonical >>= fun stats ->
-      let length = Int64.of_int(stats.Unix.st_size)
-      (* why don't they use largefile ? *)
-      in
-      Logger.info_f_ "dumping %S (%Li bytes)" canonical length >>= fun () ->
-      Lwt_io.with_file ~mode:Lwt_io.input canonical
-        (fun ic ->
-           Llio.output_int64 oc length >>= fun () ->
-           Llio.copy_stream ~length ~ic ~oc
-        )
+      let start_i = TlogMap.get_start_i tlog_map n in
+      match start_i with
+      | None -> failwith (Printf.sprintf "tlogcollection doesn't know tlog:%i" n)
+      | Some start_i ->
+         begin
+           Llio.output_string oc name >>= fun () ->
+           Sn.output_sn oc start_i    >>= fun () ->
+           File_system.stat canonical >>= fun stats ->
+           let length = Int64.of_int(stats.Unix.st_size)
+                                    (* why don't they use largefile ? *)
+           in
+           Logger.info_f_ "dumping %S (%Li bytes)" canonical length >>= fun () ->
+           Lwt_io.with_file ~mode:Lwt_io.input canonical
+                            (fun ic ->
+                              Llio.output_int64 oc length >>= fun () ->
+                              Llio.copy_stream ~length ~ic ~oc
+                            )
+           >>= fun () ->
+           Logger.debug_f_ "dumped:%s (%Li) bytes" canonical length
+         end
+
+    method private _maybe_close_current_file () =
+      match _file with
+      | None -> Lwt.return ()
+      | Some file ->
+         F.close file >>= fun () ->
+         _file <- None;
+         Lwt.return_unit
+
+    method save_tlog_file first_i remote_name length ic =
+      (* what with rotation (jump to new tlog), open streams, ...*)
+      self # _maybe_close_current_file ()
+      >>= fun () ->                                                               
+      let full_name = TlogMap.get_full_path tlog_map remote_name in
+      let extension = Tlog_map.extension_of remote_name in
+      let tmp = full_name ^ ".tmp" in
+      
+      Logger.info_f_
+        "save_tlog_file: first_i:%s remote_name:%s in %s"
+        (Sn.string_of first_i) remote_name tmp
```

Comment:
Hmm, so the tempfile name is based on the remote_name, which is kinda confusing. I think it would be clearer to use the remote name only to get the extension (see if it's a tlog/tlf/tlx), and base the temp file name on new_outer (as is done for the final filename below).